recommendation_engine/predictor/online_recommendation.py

```diff
@@ -0,0 +1,174 @@
+                  latent factors.
+        """
+        return self.VariationalAutoEncoder()
+
+    def _map_package_id_to_name(self, package_id_list):
+        """Map the package id from the model to its name.
+
+        :package_id_list: A python iterable containing all the package ids
+        """
+        return [self.package_id_name_map[str(package_id)] for package_id in package_id_list]
+
+    def _map_package_name_to_id(self, package_name_list):
+        """Map the package name to its id."""
+        return [self.package_name_id_map[package_name] for package_name in package_name_list]
+
+    def _load_package_id_to_name_map(self):
+        """Load the package-id to name mapping."""
+        self.package_id_name_map = self.s3_client.read_json_file(filename=ID_TO_PACKAGE_MAP)
+        self.package_name_id_map = self.s3_client.read_json_file(filename=PACKAGE_TO_ID_MAP)
+
+    def _find_closest_user_in_training_set(self, new_user_stack):
+        """Check if we already have the recommendations for the stack precomputed."""
+        new_user_stack = set(new_user_stack)
+        # minDiff = sys.maxsize
+        closest = None
+        for idx, stack in enumerate(self.user_stacks):
+            if stack == new_user_stack:
+                closest = idx
+                break
+            # elif len(stack.difference(new_user_stack)) < minDiff:
+            # minDiff = len(stack.difference(new_user_stack))
+            # closest = idx
+        return closest
+
+    def _sigmoid(self, x, derivative=False):
+        return x * (1 - x) if derivative else 1 / (1 + np.exp(-x))
+
+    def predict(self, new_user_stack):
+        """Predict companion packages."""
+        missing = []
+        avail = []
+        for package in new_user_stack:
+            pkg_id = self.package_name_id_map.get(package, -1)
+            if pkg_id == -1:
+                missing.append(package)
+            else:
+                avail.append(pkg_id)
+        # if more than half the packages are missing
+        if len(missing) >= len(avail):
+            return missing, []
+        new_user_stack = avail
+        # Check whether we have already seen this stack.
+        user = self._find_closest_user_in_training_set(new_user_stack)
+        if user is not None:
+            _logger.info("Have precomputed stack")
+            recommendation = np.dot(self.user_matrix[int(user), :].reshape([1, 50]),
+                                    self.latent_item_rep_mat.T)
+        else:
+            _logger.info("Calculating latent representation, have not seen this combination before")
+            scoring = PMFScoring(self.model_dict, self.item_ratings)
+            user_latent_rep = scoring.predict_transform(new_user_stack)
+            recommendation = np.dot(user_latent_rep.reshape([1, 50]),
+                                    self.latent_item_rep_mat.T)
+        packages = np.argsort(recommendation)[0][::-1].tolist()
```

Comment:
Suggestion, in the if-else condition above we can use a config parameter instead of of hard-coding the number of latent features